simulator/hosts.go

```diff
@@ -0,0 +1,159 @@
+package simulator
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+)
+
+type Host struct {
+	// Name of this host type.
+	Name string `json:"name"`
+	// parentName is the name of the parent type.
+	ParentName string `json:"parent"`
+	// parent is a pointer to the parent (nil if no parent) used for query
+	// result inheritance.
+	parent *Host
+	//QueryResults maps from regexp pattern to query results that should be
+	//returned.
+	Queries []matcher `json:"queries"`
+
+	// The following members facilitate logging unmatched queries.
+	logger           log.Logger
+	unmatchedMutex   sync.Mutex
+	unmatchedQueries map[string]bool
+}
+
+type matcher struct {
+	// Pattern is a regexp for the query patterns this should match.
+	Pattern regexp.Regexp `json:"pattern"`
+	// Results is the results to return for matched queries
+	Results []map[string]string `json:"results"`
+}
+
+type hostYAML struct {
+	Name       string `json:"name"`
+	ParentName string `json:"parent"`
+	Queries    []struct {
+		Pattern string              `json:"pattern"`
+		Results []map[string]string `json"results"`
+	} `json:"queries"`
+}
+
+func LoadHosts(dir string, logger log.Logger) (map[string]*Host, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "listing hosts directory")
+	}
+
+	hostMap := map[string]*Host{}
+
+	// Load all files
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".yaml") {
+			path := filepath.Join(dir, file.Name())
+			contents, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, errors.Wrapf(err, "reading file %s", path)
+			}
+
+			var h hostYAML
+			err = yaml.Unmarshal(contents, &h)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unmarshal yaml for %s", path)
+			}
+
+			host := Host{
+				Name:             h.Name,
+				ParentName:       h.ParentName,
+				Queries:          []matcher{},
+				unmatchedQueries: make(map[string]bool),
+				logger:           logger,
+			}
+
+			for _, q := range h.Queries {
+				re, err := regexp.Compile(strings.ToLower(q.Pattern))
+				if err != nil {
+					return nil, errors.Wrapf(err, "compile regexp for %s", path)
+				}
+				host.Queries = append(host.Queries, matcher{*re, q.Results})
+			}
+
+			// Check for duplicate host type name
+			if _, exists := hostMap[host.Name]; exists {
+				return nil, errors.Errorf("duplicate host type: %s", host.Name)
```

Comment:
Would love to see in the comment on line 89 why you want to return an error instead of logging and ignoring the duplicate.